statefun-flink/statefun-flink-core/src/main/java/org/apache/flink/statefun/flink/core/httpfn/HttpFunctionProvider.java

```diff
@@ -71,40 +68,35 @@ private HttpFunctionEndpointSpec getEndpointsSpecOrThrow(FunctionType functionTy
     throw new IllegalStateException("Unknown type: " + functionType);
   }
 
-  private RequestReplyClient buildHttpClient(
-      HttpFunctionEndpointSpec spec, FunctionType functionType) {
-    if (sharedClient == null) {
-      sharedClient = OkHttpUtils.newClient();
-    }
-    OkHttpClient.Builder clientBuilder = sharedClient.newBuilder();
-    clientBuilder.callTimeout(spec.maxRequestDuration());
-    clientBuilder.connectTimeout(spec.connectTimeout());
-    clientBuilder.readTimeout(spec.readTimeout());
-    clientBuilder.writeTimeout(spec.writeTimeout());
-
-    URI endpointUrl = spec.urlPathTemplate().apply(functionType);
+  private static RequestReplyClient buildTransportClientFromSpec(
+      URI endpointUrl, HttpFunctionEndpointSpec endpointsSpec) {
+    final RequestReplyClientFactory factory = endpointsSpec.transportClientFactory();
+    final ObjectNode properties = endpointsSpec.transportClientProperties();
 
-    final HttpUrl url;
-    if (UnixDomainHttpEndpoint.validate(endpointUrl)) {
-      UnixDomainHttpEndpoint endpoint = UnixDomainHttpEndpoint.parseFrom(endpointUrl);
-
-      url =
-          new HttpUrl.Builder()
-              .scheme("http")
-              .host("unused")
-              .addPathSegment(endpoint.pathSegment)
-              .build();
-
-      configureUnixDomainSocket(clientBuilder, endpoint.unixDomainFile);
+    if (Thread.currentThread().getContextClassLoader() == factory.getClass().getClassLoader()) {
+      // in this case, we're using one of our own shipped transport client factory
+      return factory.createTransportClient(properties, endpointUrl);
     } else {
-      url = HttpUrl.get(endpointUrl);
+      try (SetContextClassLoader ignored = new SetContextClassLoader(factory)) {
+        return new ContextSafeRequestReplyClient(
+            factory.createTransportClient(properties, endpointUrl));
+      }
     }
-    return new HttpRequestReplyClient(url, clientBuilder.build(), () -> shutdown);
   }
 
   @Override
   public void shutdown() {
-    shutdown = true;
-    OkHttpUtils.closeSilently(sharedClient);
+    specificTypeEndpointSpecs
+        .values()
+        .forEach(spec -> shutdownTransportClientFactory(spec.transportClientFactory()));
+    perNamespaceEndpointSpecs
+        .values()
+        .forEach(spec -> shutdownTransportClientFactory(spec.transportClientFactory()));
+  }
+
+  private static void shutdownTransportClientFactory(RequestReplyClientFactory factory) {
+    try (SetContextClassLoader ignored = new SetContextClassLoader(factory)) {
```

Comment:
As you can see, switching classloaders is currently a concern handled by the `HttpFunctionProvider`, and is rather error prone and fragile as we include more and more extensions in the future.

It may be worth considering moving this responsibility elsewhere, e.g. the `ExtensionResolver` should perhaps always return a classloader-safe, already decorated instance of extensions. If that's the case, we can clean up a lot of this classloader checks in the `HttpFunctionProvider`, which is essentially user code to the runtime.